### PR TITLE
CartesianProductTest × RangeSources

### LINKED ISSUE #354: Make `@CartesianProductTest` work with range sources
It seems to me that `@CartesianProductTest` may work pretty well with the range sources - e.,g.:

```java
@CartesianProductTest(name = "word: {0} / number: {1}")
@CartesianValueSource(strings = { "foo", "bar"})
@IntRangeSource(from = 0, to = 10, step = 2)
void combination(String word, int number) { }
```

What do you think, @Michael1993? Let me know of you think this is worth exploring and whether you have time for it.

### PR DESCRIPTION
**Pull request by Michael1993**
Closes #354 

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [x] There is documentation (Javadoc and site documentation; added or updated)
* [x] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [x] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [x] Only one sentence per line (especially in `.adoc` files)
* [x] Javadoc uses formal style, while sites documentation may use informal style (see #265)

Documentation (new extension)
* [x] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [x] The `package-info.java` contains information about the new extension

Code
* [x] Code adheres to code style, naming conventions etc.
* [x] Successful tests cover all changes
* [x] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks (see #164)
* [x] Tests use [AssertJ](https://joel-costigliola.github.io/assertj/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [x] A prepared commit message exists
* [x] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

**Comment by Michael1993** [comment]
Not having 100% code coverage because of "this should never happen" exceptions. (╯°□°)╯︵ ┻━┻

**Comment by Michael1993** [comment]
TODO:
* CartesianValueSource is now annotated with `@ArgumentsSource`, this might lead to weird behaviour with `@ParameterizedTest`
* Preferably all range sources should have a test
* Documentation should reflect the change

File: docs/cartesian-product.adoc
Comment (Bukama) [inline docs/cartesian-product.adoc:29]: First line has an upper `T` the other two a lower one.

File: docs/cartesian-product.adoc
Comment (Bukama) [inline docs/cartesian-product.adoc:84]: A bit below there is link to ValueSources of JUnit (can't comment that unchanged line), but the link only opens the user guide but not the anker of that sub topic

File: docs/cartesian-product.adoc
Comment (Bukama) [inline docs/cartesian-product.adoc:116]: Note: We have to check this link (and the one below) after site deployment, as of now in the PR it points to `https://github.com/junit-pioneer/junit-pioneer/blob/d0ef6d0e6d307246a54747055866f12973d9971c/docs/range-sources.adoc` but I know that it some kind of relative link.

**Review by Bukama — CHANGES_REQUESTED** [review]
As already mentioned in stream chat yesterday this looks quite good to me. Some minor things about the docs.

One maybe bigger thing we should investigate before merging this: The `@RangeSources` are now repeatable.  But I don't see any changes - not even tests - how the RangeSource itself handles them if there are multiple annotations (which is technically possible now). You wrote in the docs to not do it, but we should have investigated this a bit more to know whats happen if anyone does that.

File: docs/cartesian-product.adoc
Comment (Michael1993) [reply docs/cartesian-product.adoc:84]: https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources-ValueSource This is the link. It absolutely _should_ open the user guide at the proper anchor.

**Comment by Bukama** [comment]
HAHA I'm so stupid. I wanted to write a commit message for the PioneerAssertion, but wrote one for this 🤯 

As I already wrote it leave it here for maybe further usage:

```
Make CartesianProductTest work with RangeSources (#354 / #362)

A CartesianProduct is the set of all ordered pairs of elements from at least two sets.
The CartesianProductExtension is already able to create this set from multiple different sources, e.g. CartesianValueSource.

The RangeSourceExtension creates multiple parameters by taking an interval and a step size, e.g. create every from from 1 to 10 with step size 1.
It's available for multiple variable types, including byte, short and more.

Combining these extensions would be an easy way to provide each combination of parameters from multiple intervals.
Up to this PR the combination was not possible.
Main cause for this was that the annotations of the RangeSourceExtension were not repeatable, so it was not possible two ranges to a test method and then let the CartesianProductExtension create the CartesianProduct to use all combination of inputs from both ranges.

This PRs adds repeatable range source annotations and extends the CartesianProductExtension to build the product of the given intervals.

closes #354
PR: #362
```

**Comment by nipafx** [comment]
Couple of things that need to be done and might as well be part of this PR:

* even though [you're now a maintainer](https://github.com/junit-pioneer/junit-pioneer/commit/a07a28b024c7d24ff5aaae70ffdcb1f5fbb2ab4c), you still get to credit yourself for this extension in the README
* add `@since` annotations to the existing Cartesian product API
* add Javadoc to `CartesianProductTest.Sets`
* add a test to `CartesianProductTest.Sets#addAll(Iterable)`

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:119]:
```suggestion
This is _not_ like `@ParameterizedTest`, where you have to define all inputs in a single `@ArgumentsSource`.
```

😁

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:128]: Asciidoc source blocks and tab indentation, please. :)

File: src/main/java/org/junitpioneer/jupiter/params/ByteRangeSource.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/params/ByteRangeSource.java:40]: As these annotations are now repeatable and can be used with Cartesian products, I think they need some more changes:

* explain their dual use in their Javadoc (including a link to Cartesian product docu)
* explain where they can be repeated and where they can't

File: src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java:78]: This message may be more helpful if we explain why, even though the annotation is repeatable, it can only occur once on `@ParameterizedTests`.

File: src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java
Comment (nipafx) [inline src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java:681]: This looks great!

File: src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java
Comment (nipafx) [inline src/test/java/org/junitpioneer/jupiter/CartesianProductTestExtensionTests.java:685]: If this works with `@CartesianValueSource`, there should be a test for that.

File: src/main/java/org/junitpioneer/jupiter/PioneerAnnotationUtils.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/PioneerAnnotationUtils.java:137]: Without looking into this, I'm a but surprised that there's a combination of annotation search parameters that we don't already cover. Can you help me understand how this is different from the other methods?

File: src/main/java/org/junitpioneer/jupiter/CartesianValueArgumentsProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/CartesianValueArgumentsProvider.java:38]: Did you check newer Eclipse versions work with this?

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:121]:
```suggestion
		catch (Exception ex) {
			throw new ExtensionConfigurationException("Could not provide arguments because of exception.", ex);
		}
```

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:114]: Is `Annotation source` a `...Rangesource`? If so, can you rename the variable(s) accordingly?

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:135]: This should be...

```suggestion
		if (provider instanceof Consumer) {
```

... but, it's not? (One test fails after the change.)

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:147]: Note to self: Take a close look at this when you're more awake. :)

**Review by nipafx — COMMENTED** [review]
Looks great, I really like how this turned out. Two things:

1. Please remove curly braces around one-liners. My eyes... the googles, they do nothing.

2. We need a test that makes sure `@ParameterizedTest` fails if a (now repeatable) value source is applied twice, e.g.:

```java
@LongRangeSource(from = 1L, to = 2L)
@LongRangeSource(from = 1L, to = 5L)
@ParameterizedTest
void twoAnnotations() { }
```

**Comment by nipafx** [comment]
I've just heard, there may be a reason to introduce `CartesianAnnotationConsumer` - if there is, let's give it a shot. I like expressing ideas in types.

File: src/main/java/org/junitpioneer/jupiter/CartesianValueArgumentsProvider.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/CartesianValueArgumentsProvider.java:38]: Hm, I can't remember. I think I did, but I forgot to write it down. 🤦

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:114]: Not a `RangeSource` - it is an `Annotation` that is itself annotated with `@ArgumentsSource`.

It works like this (basically copying `ParameterizedTest`, btw):
 - Your test is a `CartesianProductTest`
 - This test has annotations on it that are annotated with `@ArgumentsSource`
 - The `@ArgumentsSource` defines and `ArgumentsProvider` (class) that must implement `AnnotationConsumer<A extends Annotation>` (in our case this should be `CartesianAnnotationConsumer`)
 - We instantiate the `ArgumentsProvider` (must have no-args constructor)
 - We call the `accept(Annotation)` method with the _original_ annotation
 - We call `provideArguments(ExtensionContext context)`

Easy-peasy, right?

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:135]: `ArgumentsProvider`s used by annotations belonging to JUnit's `@ParameterizedTest` (e.g.: `@ValueSource`) implement `Consumer` through `AnnotationConsumer`. We will fix this with defining `CartesianAnnotationConsumer`.

File: src/main/java/org/junitpioneer/jupiter/CartesianValueArgumentsProvider.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/CartesianValueArgumentsProvider.java:38]: I will revert it just to be safe - but we should check it sometime.

File: src/main/java/org/junitpioneer/jupiter/PioneerAnnotationUtils.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/PioneerAnnotationUtils.java:137]: No. 👿

Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/PioneerAnnotationUtils.java:137]: JUnits `AnnotationSupport` and our own `PioneerAnnotationUtils` has a lot of methods for getting specific annotations _directly_. We can not get annotations based on a criteria. I wanted to get all annotations on a test method that are themselves annotated with `@ArgumentsSource`. So, that's what this does - returns all annotations on an `AnnotatedElement` that are *annotated with a given annotation*.

**Comment by Bukama** [comment]
Can you give me a quick update on this. I can't see if Nicolai's requests are already in or not. Thank you.

**Comment by Michael1993** [comment]
> Can you give me a quick update on this. I can't see if Nicolai's requests are already in or not. Thank you.

Nicolai needs to review it again. I implemented the things he mentioned but it was a late-night review, so he might have missed some things earlier.

**Review by scordio — COMMENTED** [review]
Thank you @Michael1993 for your effort! I only noticed two minor points in the docs.

Looking forward to having this feature available.

File: docs/cartesian-product.adoc
Comment (Michael1993) [reply docs/cartesian-product.adoc:137]: Fixed. 👍

File: docs/cartesian-product.adoc
Comment (Michael1993) [reply docs/cartesian-product.adoc:188]: Good catch! 👍 
Fixed.

File: src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java:109]: In case this is not clear, here is an example:
```java
@ArgumentsSource(ExampleArgumentsProvider.class)
@interface ExampleSourceAnnotation {
   int intValue();
}

class ExampleArgumentsProvider implements ArgumentsProvider, CartesianAnnotationConsumer<ExampleSourceAnnotation> {
  int i;

  void accept(ExampleSourceAnnotation annotation) {
    /* do something with intValue() */
    i = annotation.intValue();
  }

  Stream<? extends Arguments> provideArguments(ExtensionContext context) {
    /* `accept` was already called so `i` should have a value */
    return Stream.of(Arguments.of(i));
  }
}
```

So, if we have a `CartesianProductTest`, annotated with `ExampleSourceAnnotation`, we store the annotation, instantiate `ExampleArgumentsProvider` (with no-arg constructor), 'feed it' the `ExampleSourceAnnotation` and call `provideArguments` to get the set for our test.

File: src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java:73]: Will this find meta-annotations? I doubt it would. Does this make sense?

```java
@IntRangeSource(from = 0, to = 10)
public @interface FromZeroToTen { }

@CartesianProductTest
@FromZeroToTen
void test(int i) {
}
```

If it does, we need to make sure that we find such annotations. Please use/extend `org.junitpioneer.jupiter.params.PioneerAnnotationUtils` if it's suitable - as painful as that may be.

File: src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java:104]: I just saw this code in `PioneerAnnotationUtils`, no need to repeat it here. Just expose there and add (reflectively) to the `params` variant.

File: docs/cartesian-product.adoc
Comment (nipafx) [inline docs/cartesian-product.adoc:168]: Please document how to create `ArgumentsProvider` (as promised in `CartesianAnnotationConsumer`).

**Review by nipafx — COMMENTED** [review]
Looking good, we're getting there.

**Review by Bukama — APPROVED** [review]
I like the meaningfull names etc (and I don't like if-clauses without brackets, but I had this conversion with Nicolai some time ago, too...)

I keep my state that this looks good for me. I leave final thoughts to Nicolai

File: src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java:73]: All range sources have `@Target(ElementType.METHOD)` on them. No meta-annotations can exist.

Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/params/RangeSourceArgumentsProvider.java]: True. We should probably discuss on an issue first whether this makes sense.

**Comment by nipafx** [comment]
Looks good to me, merge ahead! We still need a commit message, though, and I'll cowardly leave this to somebody else. 😊

**Comment by nipafx** [comment]
Never mind, found @Bukama's. :)

**Pull request merged**
